service/history/queues/executable_dlq.go

```diff
@@ -0,0 +1,107 @@
+	"go.temporal.io/server/common/persistence/serialization"
+)
+
+type (
+	// ExecutableDLQ is an Executable decorator that will enqueue the task to a DLQ if the underlying Executable fails.
+	// The first call to Execute for which the underlying Executable returns a terminal error will return an
+	// ErrTerminalTaskFailure error and change the behavior of this task to just try to send itself to the DLQ.
+	// Specifically, all subsequent calls to Execute will attempt to send the task to the DLQ. If this fails, the error
+	// will be returned (with the expectation that clients will retry). If it succeeds, no error will be returned and
+	// the task can be acked. When the executable is in the failed state, all calls to HandleErr will just return the
+	// error passed in.
+	// TODO: wrap all executables with this
+	// TODO: add metrics and logging to this
+	ExecutableDLQ struct {
+		Executable
+		dlq DLQ
+		// dlqCause is the original error which caused this task to be sent to the DLQ. It is only set once.
+		dlqCause        error
+		clusterMetadata ClusterMetadata
+	}
+
+	// DLQ is a dead letter queue that can be used to enqueue tasks that fail to be processed.
+	DLQ interface {
+		EnqueueTask(
+			ctx context.Context,
+			request *persistence.EnqueueTaskRequest,
+		) (*persistence.EnqueueTaskResponse, error)
+	}
+
+	// ClusterMetadata is an interface that provides access to cluster metadata.
+	ClusterMetadata interface {
+		GetCurrentClusterName() string
+	}
+)
+
+var (
+	_                      Executable = new(ExecutableDLQ)
+	ErrTerminalTaskFailure            = errors.New("original task failed and this task is now to send the original to the DLQ")
+	ErrSendTaskToDLQ                  = errors.New("failed to send task to DLQ")
+)
+
+// NewExecutableDLQ wraps an Executable to ensure that it is sent to the DLQ if it fails terminally.
+func NewExecutableDLQ(dlq DLQ, executable Executable, clusterMetadata ClusterMetadata) *ExecutableDLQ {
+	return &ExecutableDLQ{
+		Executable:      executable,
+		dlq:             dlq,
+		clusterMetadata: clusterMetadata,
+	}
+}
+
+// Execute is not thread-safe.
+func (d *ExecutableDLQ) Execute() error {
+	if d.dlqCause == nil {
+		// This task has not experienced a terminal failure yet, so we should execute it.
+		err := d.Executable.Execute()
+		if !errors.As(err, new(*serialization.DeserializationError)) &&
+			!errors.As(err, new(*serialization.UnknownEncodingTypeError)) {
+			return err
+		}
+		d.dlqCause = err
+		return fmt.Errorf("%w: %v", ErrTerminalTaskFailure, err)
+	}
+	// This task experienced a terminal failure, so we should try to send it to the DLQ.
+	clusterName := d.clusterMetadata.GetCurrentClusterName()
+	_, err := d.dlq.EnqueueTask(context.TODO(), &persistence.EnqueueTaskRequest{
+		QueueType:     persistence.QueueTypeHistoryDLQ,
+		SourceCluster: clusterName,
+		TargetCluster: clusterName,
+		Task:          d.GetTask(),
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrSendTaskToDLQ, err)
```

Comment:
hmmm, we will need the HandlerErr() in executable to handle wrapper errors so that proper retry (reschedule) policy can be used. 

I have a PR for that before but never have that landed. I will follow up in next 1-2 weeks, but it won't block your DLQ works.

Reply:
Sounds good. I'll keep that in mind.